app/src/main/java/lt/gyvosistorijos/service/GeofenceIntentService.kt

```diff
@@ -47,68 +54,87 @@ class GeofenceIntentService : IntentService(GeofenceIntentService.TAG) {
+        Timber.i("Geofence triggered at $triggeredLocation with transition $transition: $triggeringGeofencesIdsString")
     }
 
-    /**
-     * Posts a notification in the notification bar when a transition is detected.
-     * If the user clicks the notification, control goes to the MainActivity.
-     */
-    private fun sendNotification() {
-        // Create an explicit content Intent that starts the main Activity.
+    // Picasso holds Target instance with weak reference.
+    // So it is better to hold Target as instance field.
+    // see: http://stackoverflow.com/a/29274669/5183999
+    private var notificationTarget: Target? = null
+
+    private fun sendNotification(story: Story) {
         val notificationIntent = Intent(applicationContext, MainActivity::class.java)
 
-        // Construct a task stack.
         val stackBuilder = TaskStackBuilder.create(this)
 
-        // Add the main Activity to the task stack as the parent.
         stackBuilder.addParentStack(MainActivity::class.java)
 
-        // Push the content Intent onto the stack.
         stackBuilder.addNextIntent(notificationIntent)
 
-        // Get a PendingIntent containing the entire back stack.
         val notificationPendingIntent = stackBuilder.getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT)
 
-        // Get a notification builder that's compatible with platform versions >= 4
         val builder = NotificationCompat.Builder(this)
 
-        // Define the notification settings.
         builder.setSmallIcon(R.mipmap.ic_launcher)
                 .setContentTitle(getString(R.string.geofence_notification_text))
-                .setContentIntent(notificationPendingIntent).priority = NotificationCompat.PRIORITY_MIN
+                .setContentText(story.text)
+                .setCategory(NotificationCompat.CATEGORY_RECOMMENDATION)
+                .setContentIntent(notificationPendingIntent).priority = NotificationCompat.PRIORITY_LOW
 
-        // Get an instance of the Notification manager
         val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
 
-        // Issue the notification
-        notificationManager.notify(0, builder.build())
+        notificationTarget = object : Target {
+            override fun onBitmapFailed(errorDrawable: Drawable?) {
+                notificationManager.notify(0, builder.build())
+            }
+
+            override fun onBitmapLoaded(bitmap: Bitmap?, from: Picasso.LoadedFrom?) {
+                if (bitmap != null) {
+                    builder.setLargeIcon(bitmap)
+
+                    notificationManager.notify(0, builder.build())
+                } else {
+                    Timber.d("Notification loaded bitmap is null")
+                }
+            }
+
+            override fun onPrepareLoad(placeHolderDrawable: Drawable?) {
+            }
+
+        }
+
+        Handler(mainLooper).post({
+            Picasso.with(this)
```

Comment:
I thought that it is. Although Picasso has check from which thread it was called.
```java
  static void checkMain() {
    if (!isMain()) {
      throw new IllegalStateException("Method call should happen from the main thread.");
    }
  }
```

When starting from another thread:
```
java.lang.IllegalStateException: Method call should happen from the main thread.
                                                                              at com.squareup.picasso.Utils.checkMain(Utils.java:136)
                                                                              at com.squareup.picasso.RequestCreator.into(RequestCreator.java:496)
                                                                              at lt.gyvosistorijos.service.GeofenceIntentService.sendNotification(GeofenceIntentService.kt:134)
                                                                              at lt.gyvosistorijos.service.GeofenceIntentService.onHandleIntent(GeofenceIntentService.kt:69)
                                                                              at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:68)
                                                                              at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                              at android.os.Looper.loop(Looper.java:154)
                                                                              at android.os.HandlerThread.run(HandlerThread.java:61)
```